File: mediapipe/java/com/google/mediapipe/glutil/ShaderUtil.java
// Copyright 2019 The MediaPipe Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.mediapipe.glutil;

import android.graphics.Bitmap;
import android.opengl.GLES20;
import android.opengl.GLUtils;
import com.google.common.flogger.FluentLogger;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.FloatBuffer;
import java.util.Map;
import javax.annotation.Nullable;

/**
 * Utility class for managing GLSL shaders.
 */
public class ShaderUtil {
  private static final FluentLogger logger = FluentLogger.forEnclosingClass();

  /**
   * Loads a shader from source.
   * @param shaderType a valid GL shader type, e.g. {@link GLES20#GL_VERTEX_SHADER} or
   *        {@link GLES20#GL_FRAGMENT_SHADER}.
   * @param source the shader's source in text form.
   * @return a handle to the created shader, or 0 in case of error.
   */
  public static int loadShader(int shaderType, String source) {
    int shader = GLES20.glCreateShader(shaderType);
    GLES20.glShaderSource(shader, source);
    GLES20.glCompileShader(shader);
    int[] compiled = new int[1];
    GLES20.glGetShaderiv(shader, GLES20.GL_COMPILE_STATUS, compiled, 0);
    if (compiled[0] == 0) {
      logger.atSevere().log("Could not compile shader %d: %s", shaderType,
          GLES20.glGetShaderInfoLog(shader));
      GLES20.glDeleteShader(shader);
      shader = 0;
    }
    return shader;
  }

  /**
   * Creates a shader program.
   *
   * @param vertexSource source of the vertex shader.
   * @param fragmentSource source of the fragment shader.
   * @param attributeLocations a map of desired locations for attributes. Can be null.
   * @return a handle to the created program, or 0 in case of error.
   */
  public static int createProgram(
      String vertexSource,
      String fragmentSource,
      @Nullable Map<String, Integer> attributeLocations) {
    int vertexShader = loadShader(GLES20.GL_VERTEX_SHADER, vertexSource);
    if (vertexShader == 0) {
      return 0;
    }
    int fragmentShader = loadShader(GLES20.GL_FRAGMENT_SHADER, fragmentSource);
    if (fragmentShader == 0) {
      return 0;
    }

    int program = GLES20.glCreateProgram();
    if (program == 0) {
      logger.atSevere().log("Could not create program");
    }
    GLES20.glAttachShader(program, vertexShader);
    GLES20.glAttachShader(program, fragmentShader);

    if (attributeLocations != null) {
      for (Map.Entry<String, Integer> entry : attributeLocations.entrySet()) {
        GLES20.glBindAttribLocation(program, entry.getValue(), entry.getKey());
      }
    }

    GLES20.glLinkProgram(program);
    int[] linkStatus = new int[1];
    GLES20.glGetProgramiv(program, GLES20.GL_LINK_STATUS, linkStatus, 0);
    if (linkStatus[0] != GLES20.GL_TRUE) {
      logger.atSevere().log("Could not link program: %s", GLES20.glGetProgramInfoLog(program));
      GLES20.glDeleteProgram(program);
      program = 0;
    }
    return program;
  }

  /**
   * Creates a texture. Binds it to texture unit 0 to perform setup.
   * @return the name of the new texture.
   */
  public static int createRgbaTexture(int width, int height) {
    final int[] textureName = new int[] {0};
    GLES20.glGenTextures(1, textureName, 0);

    GLES20.glActiveTexture(GLES20.GL_TEXTURE0);
    GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, textureName[0]);
    GLES20.glTexImage2D(
        GLES20.GL_TEXTURE_2D,
        0,
        GLES20.GL_RGBA,
        width, height,
        0,
        GLES20.GL_RGBA,
        GLES20.GL_UNSIGNED_BYTE,
        null);
    ShaderUtil.checkGlError("glTexImage2D");
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_MIN_FILTER, GLES20.GL_LINEAR);
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_MAG_FILTER, GLES20.GL_LINEAR);
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_WRAP_S, GLES20.GL_CLAMP_TO_EDGE);
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_WRAP_T, GLES20.GL_CLAMP_TO_EDGE);
    ShaderUtil.checkGlError("texture setup");
    return textureName[0];
  }

  /**
   * Creates a texture from a Bitmap. Binds it to texture unit 0 to perform setup.
   *
   * @return the name of the new texture.
   */
  public static int createRgbaTexture(Bitmap bitmap) {
    final int[] textureName = new int[] {0};
    GLES20.glGenTextures(1, textureName, 0);

    GLES20.glActiveTexture(GLES20.GL_TEXTURE0);
    GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, textureName[0]);
    GLUtils.texImage2D(GLES20.GL_TEXTURE_2D, 0, bitmap, 0);
    ShaderUtil.checkGlError("texImage2D");
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_MIN_FILTER, GLES20.GL_LINEAR);
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_MAG_FILTER, GLES20.GL_LINEAR);
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_WRAP_S, GLES20.GL_CLAMP_TO_EDGE);
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_WRAP_T, GLES20.GL_CLAMP_TO_EDGE);
    ShaderUtil.checkGlError("texture setup");
    return textureName[0];
  }

  /**
   * Creates a {@link FloatBuffer} with the given arguments as contents.
   * The buffer is created in native format for efficient use with OpenGL.
   */
  public static FloatBuffer floatBuffer(float... values) {
    ByteBuffer byteBuffer =
        ByteBuffer.allocateDirect(
            values.length * 4 /* sizeof(float) */);
    // use the device hardware's native byte order
    byteBuffer.order(ByteOrder.nativeOrder());

    // create a floating point buffer from the ByteBuffer
    FloatBuffer floatBuffer = byteBuffer.asFloatBuffer();
    // add the coordinates to the FloatBuffer
    floatBuffer.put(values);
    // set the buffer to read the first coordinate
    floatBuffer.position(0);
    return floatBuffer;
  }

  /**
   * Calls {@link GLES20#glGetError} and raises an exception if there was an error.
   */
  public static void checkGlError(String msg) {
    int error = GLES20.glGetError();
    if (error != GLES20.GL_NO_ERROR) {
      throw new RuntimeException(msg + ": GL error: 0x" + Integer.toHexString(error));
    }
  }
}

File: mediapipe/gpu/shader_util.cc
// Copyright 2019 The MediaPipe Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "mediapipe/gpu/shader_util.h"

#include <stdlib.h>

#include <cmath>
#include <string>
#include <vector>

#include "absl/log/absl_check.h"
#include "absl/log/absl_log.h"
#include "absl/strings/str_format.h"
#include "absl/strings/str_join.h"
#include "absl/strings/str_split.h"
#include "mediapipe/framework/port/logging.h"

#if DEBUG
#define GL_DEBUG_LOG(type, object, action)                        \
  do {                                                            \
    GLint log_length = 0;                                         \
    glGet##type##iv(object, GL_INFO_LOG_LENGTH, &log_length);     \
    if (log_length > 0) {                                         \
      GLchar* log = static_cast<GLchar*>(malloc(log_length));     \
      glGet##type##InfoLog(object, log_length, &log_length, log); \
      ABSL_LOG(INFO) << #type " " action " log:\n" << log;        \
      free(log);                                                  \
    }                                                             \
  } while (0)
#else
#define GL_DEBUG_LOG(type, object, action)
#endif

#define GL_ERROR_LOG(type, object, action)                        \
  do {                                                            \
    GLint log_length = 0;                                         \
    glGet##type##iv(object, GL_INFO_LOG_LENGTH, &log_length);     \
    if (log_length > 0) {                                         \
      GLchar* log = static_cast<GLchar*>(malloc(log_length));     \
      glGet##type##InfoLog(object, log_length, &log_length, log); \
      ABSL_LOG(ERROR) << #type " " action " log:\n" << log;       \
      free(log);                                                  \
    }                                                             \
  } while (0)

namespace mediapipe {
namespace {

constexpr int kMaxShaderInfoLength = 1024;

std::string AddLineNumbers(const GLchar* source) {
  // Use format "%ni %s", with n=1 for 1..9 lines, n=2 for 10..99 lines etc.
  // Note that StrFormat needs either a constexpr format or a ParsedFormat.
  std::vector<std::string> lines = absl::StrSplit(source, '\n');
  std::string format = absl::StrFormat(
      "%%%ii %%s", static_cast<int>(ceilf(log10(1 + lines.size()))));
  auto parsed_format = absl::ParsedFormat<'i', 's'>::New(format);
  ABSL_CHECK(parsed_format);
  for (int n = 0; n < lines.size(); n++) {
    lines[n] = absl::StrFormat(*parsed_format, n + 1, lines[n]);
  }
  return absl::StrJoin(lines, "\n");
}

}  // namespace

GLint GlhCompileShader(GLenum target, const GLchar* source, GLuint* shader,
                       bool force_log_errors) {
  *shader = glCreateShader(target);
  if (*shader == 0) {
    return GL_FALSE;
  }
  glShaderSource(*shader, 1, &source, NULL);
  glCompileShader(*shader);

  GL_DEBUG_LOG(Shader, *shader, "compile");

#if UNSAFE_EMSCRIPTEN_SKIP_GL_ERROR_HANDLING
  if (!force_log_errors) {
    return GL_TRUE;
  }
#endif  // UNSAFE_EMSCRIPTEN_SKIP_GL_ERROR_HANDLING

  GLint status;

  glGetShaderiv(*shader, GL_COMPILE_STATUS, &status);
  ABSL_LOG_IF(ERROR, status == GL_FALSE) << "Failed to compile shader:\n"
                                         << AddLineNumbers(source);

  if (status == GL_FALSE) {
    int length = 0;
    GLchar cmessage[kMaxShaderInfoLength];
    glGetShaderInfoLog(*shader, kMaxShaderInfoLength, &length, cmessage);
    ABSL_LOG(ERROR) << "Error message: " << std::string(cmessage, length);
  }
  return status;
}

GLint GlhLinkProgram(GLuint program, bool force_log_errors) {
  glLinkProgram(program);

#if UNSAFE_EMSCRIPTEN_SKIP_GL_ERROR_HANDLING
  if (!force_log_errors) {
    return GL_TRUE;
  }
#endif  // UNSAFE_EMSCRIPTEN_SKIP_GL_ERROR_HANDLING

  GLint status;

  GL_DEBUG_LOG(Program, program, "link");

  glGetProgramiv(program, GL_LINK_STATUS, &status);
  ABSL_LOG_IF(ERROR, status == GL_FALSE)
      << "Failed to link program " << program;

  return status;
}

GLint GlhValidateProgram(GLuint program) {
  GLint status;

  glValidateProgram(program);

  GL_DEBUG_LOG(Program, program, "validate");

  glGetProgramiv(program, GL_VALIDATE_STATUS, &status);
  ABSL_LOG_IF(ERROR, status == GL_FALSE)
      << "Failed to validate program " << program;

  return status;
}

GLint GlhCreateProgram(const GLchar* vert_src, const GLchar* frag_src,
                       GLsizei attr_count, const GLchar* const* attr_names,
                       const GLint* attr_locations, GLuint* program,
                       bool force_log_errors) {
  GLuint vert_shader = 0;
  GLuint frag_shader = 0;
  GLint ok = GL_TRUE;

  *program = glCreateProgram();
  if (*program == 0) {
    return GL_FALSE;
  }

  ok = ok && GlhCompileShader(GL_VERTEX_SHADER, vert_src, &vert_shader,
                              force_log_errors);
  ok = ok && GlhCompileShader(GL_FRAGMENT_SHADER, frag_src, &frag_shader,
                              force_log_errors);

  if (ok) {
    glAttachShader(*program, vert_shader);
    glAttachShader(*program, frag_shader);

    // Attribute location binding must be set before linking.
    for (int i = 0; i < attr_count; i++) {
      glBindAttribLocation(*program, attr_locations[i], attr_names[i]);
    }

    ok = GlhLinkProgram(*program, force_log_errors);

    glDetachShader(*program, frag_shader);
    glDetachShader(*program, vert_shader);
  }

  if (vert_shader) glDeleteShader(vert_shader);
  if (frag_shader) glDeleteShader(frag_shader);

  if (!ok) {
    glDeleteProgram(*program);
    *program = 0;
  }

  return ok;
}

bool CompileShader(GLenum shader_type, const std::string& shader_source,
                   GLuint* shader) {
  *shader = glCreateShader(shader_type);
  if (*shader == 0) {
    VLOG(2) << "Unable to create shader of type: " << shader_type;
    return false;
  }
  const char* shader_source_cstr = shader_source.c_str();
  glShaderSource(*shader, 1, &shader_source_cstr, NULL);
  glCompileShader(*shader);

  GLint compiled;
  glGetShaderiv(*shader, GL_COMPILE_STATUS, &compiled);
  if (!compiled) {
    VLOG(2) << "Unable to compile shader:\n"
            << AddLineNumbers(shader_source_cstr);
    GL_ERROR_LOG(Shader, *shader, "compile");
    glDeleteShader(*shader);
    *shader = 0;
    return false;
  }
  return true;
}

bool CreateShaderProgram(
    GLuint vertex_shader, GLuint fragment_shader,
    const std::unordered_map<GLuint, std::string>& attributes,
    GLuint* shader_program) {
  *shader_program = glCreateProgram();
  if (*shader_program == 0) {
    VLOG(2) << "Unable to create shader program";
    return false;
  }
  glAttachShader(*shader_program, vertex_shader);
  glAttachShader(*shader_program, fragment_shader);

  for (const auto& it : attributes) {
    glBindAttribLocation(*shader_program, it.first, it.second.c_str());
  }
  glLinkProgram(*shader_program);

  GLint is_linked = 0;
  glGetProgramiv(*shader_program, GL_LINK_STATUS, &is_linked);
  if (!is_linked) {
    glDeleteProgram(*shader_program);
    *shader_program = 0;
    return false;
  }
  return true;
}

}  // namespace mediapipe
